Common/src/Auth/Authentication/Authenticator.cs

```diff
@@ -177,54 +185,70 @@ protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
   }
 
   public async Task<UserIdentity?> GetIdentityFromCertificateAsync(string cn,
-                                                              string fingerprint)
+                                                                   string fingerprint)
   {
     logger_.LogDebug("Authenticating request with CN {CN} and fingerprint {Fingerprint}",
                      cn,
                      fingerprint);
     var result = await authTable_.GetIdentityAsync(cn,
-                                                   fingerprint,
-                                                   new CancellationToken(false))
+                                                   fingerprint)
                                  .ConfigureAwait(false);
 
     if (result == null)
+    {
       return null;
+    }
     return new UserIdentity(result,
                             SchemeName);
   }
 
-  public async Task<UserIdentity?> GetImpersonationatedIdentity(UserIdentity baseIdentity,
-                                                                string       impersonnated)
+  public async Task<(UserIdentity?, string)> GetImpersonatedIdentityAsync(UserIdentity baseIdentity,
+                                                                          string?      impersonationId,
+                                                                          string?      impersonationUsername)
   {
     // Get all roles that can be impersonated
     var impersonatableRoles = baseIdentity.Claims.Where(c => c.Type == Permissions.Impersonate.Claim.Type)
                                           .Select(c => c.Value);
-    UserAuthenticationResult? result;
-    // Impersonation can be done using the username or the userid
-    if (impersonationWithUsername_)
+    UserAuthenticationResult? result = null;
+    if (impersonationId != null)
```

Comment:
Couldn't you write a single request to handle both cases, and thus avoid to wait for the first request to finish before doing the second one?